hathor/utils/yaml.py

```diff
@@ -11,13 +11,57 @@
 #  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 #  See the License for the specific language governing permissions and
 #  limitations under the License.
-
-from typing import Any, Dict
+import os
+from pathlib import Path
+from typing import Any, Union
 
 import yaml
 
+from hathor.utils.dict import deep_merge
+
+_EXTENDS_KEY = 'extends'
 
-def dict_from(*, filepath: str) -> Dict[str, Any]:
+
+def dict_from_yaml(*, filepath: Union[Path, str]) -> dict[str, Any]:
     """Takes a filepath to a yaml file and returns a dictionary with its contents."""
+    if not os.path.isfile(filepath):
+        raise ValueError(f"'{filepath}' is not a file")
+
     with open(filepath, 'r') as file:
-        return yaml.safe_load(file)
+        contents = yaml.safe_load(file)
+
+        if contents is None:
+            return {}
+
+        if not isinstance(contents, dict):
+            raise ValueError(f"'{filepath}' cannot be parsed as a dictionary")
+
+        return contents
+
+
+def dict_from_extended_yaml(*, filepath: Union[Path, str]) -> dict[str, Any]:
+    """
+    Takes a filepath to a yaml file and returns a dictionary with its contents.
+    Supports extending another yaml file via the 'extends' key in the file.
+
+    Note: the 'extends' key is reserved and will not be present in the returned dictionary.
+    To opt-out of the extension feature, use dict_from_yaml().
+    """
+    extension_dict = dict_from_yaml(filepath=filepath)
+    base_file = extension_dict.pop(_EXTENDS_KEY, None)
+
+    if not base_file:
+        return extension_dict
+
+    root_path = Path(filepath).parent
```

Comment:
Would it be simple to support always considering `os.path.dirname(hathor.conf.__file__)` as a possible base? For me it would be really useful to be able to do `extend: mainnet.yml` from outside the repo without having to explicitly put the path of the repo.

For me it is useful because I have 4 working trees of the code (`hathor-core`, `hathor-core2`, `hathor-core3`, `hathor-core4`) and 3 local nodes (but I can add more) that aren't tied to any work trees, so I can go into `hathor-core2` and start my node1 or node3 or both, each node has its own config and env vars, but I'd like that the config can extend mainnet config without having to use a full path to it since in theory it can point to any work tree, it just depends on where I launch the node from. With the previous module settings system I can just do this:

```python
from hathor.conf import mainnet as base

SETTINGS = base.SETTINGS._replace(
    # BOOTSTRAP_DNS=[],
    # MAX_PEER_CONNECTION_ATTEMPS=math.inf,
)
```

With the new system, I'd like to be able to use:

```yaml
extends: mainnet.yml
#extends: ../../hathor-core/conf/mainnet.yml  # <- instead of this
#BOOTSTRAP_DNS: []
#MAX_PEER_CONNECTION_ATTEMPS: 999999
```

Reply:
Or, an alternative to having this "lookup path", would be using internal names to point to internal configs, like:

```yaml
extends: __mainnet__ # is replaced by `Path(hathor.conf.__file__).parent / 'mainnet.yml'`
#BOOTSTRAP_DNS: []
#MAX_PEER_CONNECTION_ATTEMPS: 999999
```

Reply:
I added support for extending paths relative to `conf`: https://github.com/HathorNetwork/hathor-core/pull/618/commits/9a41c03e8b5e42fad57aa93363e3267d37d65007

Actually, I added a new `custom_root` arg to keep it generic, and `HathorSettings` uses it, setting `conf` as a custom root.